scripts/publish-npm.js

```diff
@@ -0,0 +1,121 @@
+const { execSync } = require('child_process');
+const fs = require('fs');
+const path = require('path');
+const assert = require('assert');
+const readline = require('readline');
+
+const rootPath = path.resolve('__dirname', '..');
+
+function requireSegmentApiKey() {
+  const MONGOSH_SEGMENT_API_KEY = process.env.MONGOSH_SEGMENT_API_KEY || '';
+
+  if (!MONGOSH_SEGMENT_API_KEY) {
+    throw new Error('MONGOSH_SEGMENT_API_KEY environment variable must be set');
+  }
+
+  if (MONGOSH_SEGMENT_API_KEY.length !== 32) {
+    throw new Error('MONGOSH_SEGMENT_API_KEY does not seem to have the proper format');
+  }
+
+  return MONGOSH_SEGMENT_API_KEY;
+}
+
+function gitClone(repo, dest) {
+  return execSync(`git clone ${repo} ${dest}`);
+}
+
+function getGitRemoteUrl() {
+  return execSync('git config --get remote.origin.url', {cwd: rootPath}).toString().trim();
+}
+
+function getGitRemoteHeadSHA(remoteUrl) {
+  return execSync(
+    `git ls-remote ${remoteUrl} HEAD`, {cwd: rootPath}
+  ).toString().split(/\s+/)[0].trim();
+}
+
+function writeSegmentApiKey(segmentApiKey, releaseDirPath) {
+  const analyticsConfigPath = path.resolve(
+    releaseDirPath,
+    'packages',
+    'cli-repl',
+    'lib',
+    'analytics-config.js'
+  );
+
+  console.info(`writing ${segmentApiKey} to ${analyticsConfigPath}`);
+
+  fs.writeFileSync(
+    analyticsConfigPath,
+    `module.exports = { SEGMENT_API_KEY: '${segmentApiKey}' };`
+  );
+
+  assert.equal(require(analyticsConfigPath).SEGMENT_API_KEY, segmentApiKey);
+}
+
+function isDone(task, releaseDirPath) {
+  return fs.existsSync(`${releaseDirPath}-${task}.done`);
+}
+
+function done(task, releaseDirPath) {
+  fs.writeFileSync(`${releaseDirPath}-${task}.done`, '');
+}
+
+function publish() {
+  const segmentApiKey = requireSegmentApiKey();
+  const remoteUrl = getGitRemoteUrl();
+  const remoteHeadSha = getGitRemoteHeadSHA(remoteUrl);
+  const releaseDirPath = path.resolve(rootPath, 'tmp', 'releases', remoteHeadSha);
+
+  if (!isDone('clone', releaseDirPath)) {
+    console.info(`cloning '${remoteUrl}' to '${releaseDirPath}'`);
+    gitClone(remoteUrl, releaseDirPath);
+    done('clone', releaseDirPath);
```

Comment:
I keep reading this as a `callback` "done" and keep not seeing the callback 😂 😂 

The `done` function is meant to be a log of what was finished as part of this script? Maybe we can call this function `log`?

Reply:
`done` marks the task as done (writing a file), so running `npm publish` twice would skip what already succeeded. We could do this instead:

```
doOnce('clone', releaseDirPath, () => {
    console.info(`cloning '${remoteUrl}' to '${releaseDirPath}'`);
    gitClone(remoteUrl, releaseDirPath);
});
```